ocaml/typing/jkind.mli

```diff
@@ -455,27 +480,36 @@ type sub_or_intersect =
 
 (** [sub_or_intersect t1 t2] does a subtype check, returning a [sub_or_intersect];
     see comments there for more info. *)
-val sub_or_intersect : t -> t -> sub_or_intersect
+val sub_or_intersect : l -> r -> sub_or_intersect
 
 (** [sub_or_error t1 t2] does a subtype check, returning an appropriate
     [Violation.t] upon failure. *)
-val sub_or_error : t -> t -> (unit, Violation.t) result
+val sub_or_error : l -> r -> (unit, Violation.t) result
+
+(** Like [sub], but returns the subjkind with an updated history.
+    Pre-condition: the super jkind must be fully settled; no variables
+    which might be filled in later. *)
+val sub_jkind_l : l -> l -> (l, Violation.t) result
+
+(* CR layouts v2.8: This almost certainly has to get rewritten, as l-kinds do
+   not support meets. *)
 
-(** Like [sub], but returns the subjkind with an updated history. *)
-val sub_with_history : t -> t -> (t, Violation.t) result
+(** Like [intersection_or_error], but between an [l] and an [l], as an [l]. *)
+val intersect_l_l :
+  reason:History.interact_reason -> l -> l -> (l, Violation.t) result
 
 (** Checks to see whether a jkind is the maximum jkind. Never does any
     mutation. *)
-val is_max : t -> bool
+val is_max : r -> bool
 
 (** Checks to see whether a jkind has layout any. Never does any mutation. *)
-val has_layout_any : t -> bool
+val has_layout_any : 'd t -> bool
```

Comment:
Yes, but the layout bit -- which is all this cares about -- doesn't interact with the l/r bit. In practice, we use this both on l-kinds and r-kinds. But because the layouts are separate, it's all OK.

Reply:
This surprises me a little. The layouts do have an ordering, and I’d expect that ordering to also be factored into the `l` and `r` distinction. In particular, when we add higher-kinds we will gain some `l` layouts so it will need to be this way at some point.

Reply:
OK. I've fixed this now.